### ISSUE TITLE
Map and Set not supported in v-for directive

### ISSUE THREAD

**Issue by sgreene** [susanBuck/e28-spring21#15]
**Describe the problem**
Data structures like Map and Set are pretty standard for Java developers like myself and I thought either of these data sets would be a more natural fit for the shopping list question (quantity being a key of the map and item being the value). However while Javascript ES6 does have support for these datastructures and Vue.js doesn't complain about them, I found that iterating a Map (and Set) in Vue.js [isn't supported](https://github.com/vuejs/vue/issues/6644). There are workarounds but they're sort of awkward. I guess that's why Susan suggested using multidimensional arrays! I should have taken the hint!

**To Reproduce**
Try iterating a Map object with v-for:
```js
data() {
        return {
            items: new Map(),
            qty: 0,
            item: "Oranges"
        }
    },
```
Vue.js does not complain about the above. However there's no way to easily iterate over the items. Something like this does not exist:
```html
<li v-for='item in items.keys()'/>
```

**Comment by susanBuck** [comment]
I don't take advantage of Map and Set as much as I should, so I hadn't encountered this.

Digging through the thread you linked to, I followed a link to another thread and it looks like they did add support for Map/Set in Vue 3: https://github.com/vuejs/vue/issues/2410#issuecomment-434990853

Digging through the [API docs for v-for](https://v3.vuejs.org/api/directives.html#v-for), it confirms this:

> v-for can also work on values that implement the [Iterable Protocol](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#The_iterable_protocol), including native Map and Set.

I worked up a small example to test it, and it did work: https://jsfiddle.net/0qfwxy45/3/

Curious where you hit a block with this in your tests. All the code you shared in your post looks valid - of course your map is empty, so I wouldn't expect any results to display - but assuming you had some other code that was adding to the Map?

Let me know if I'm missing something. 

Otherwise, thanks for sharing this. I enjoyed digging into the thread you shared and following the debate about needing support for Map/Set. Always interesting to me to see the "inner workings" of how framework features evolve. : )

**Comment by sgreene** [comment]
Thanks for the example! I think I just gave up on this too soon and when I saw the Github issue I linked to I just switched back to the array format. I was running out of time to submit the assignment anyway. Even more embarrassing was the code I posted... I hadn't tried it! I got stuck because in my original code I was trying to call `next()` on the `getKeys()` method of the map which is unnecessary (and caused all sorts of problems) based on the Iterable Protocol (which I did not know about until reading your message). So had I actually tried the example code I posted:
```html
<li v-for='item in items.keys()'/>
```
It would have worked! (I was populating the Map correctly and I could tell from the console debug). Sigh.. hopefully I didn't mislead anyone.
